### DISCUSSION TITLE
How to access variable from Julia chunk in R

### DISCUSSION THREAD

**Discussion by DominiqueMakowski** [quarto-dev/quarto-cli#5607] [Q&A]
Imagine I have these two chunks:

```{julia}
x = 3
```

```{r}
x
```

How can I access the variables in the Julia workspace? I know the [JuliaCall](https://cran.r-project.org/web/packages/JuliaCall/readme/README.html) package allows for some inter-operability, but here I don't want to execute a Julia command in R, rather access a variable already created in the Julia workspace. Thanks!

Note: Markdown doesn't render but the first chunk is defined as ```{julia} ... and the second as \`\`\`{r} ...

**Comment by mcanouil** [comment]
See #3792 and mostly #3701 and #4303

**Comment by mcanouil** [comment]
Closing as a duplicate and in favour of #3701 or #4303

**Comment by DominiqueMakowski** [comment]
Thanks @mcanouil Unfortunately the proposed solutions don't seem to work:

If I do:

\```{julia}
x = 3
\```


\```{r}
JuliaCall::julia_eval("x")
\```

it throws:

```
Error: Error happens in Julia.
UndefVarError: `x` not defined
Stacktrace:
 [1] top-level scope
   @ :0
 [2] eval
   @ .\boot.jl:370 [inlined]
 [3] eval_string(x::String)
   @ Main.JuliaCall C:\Users\domma\AppData\Local\R\win-library\4.2\JuliaCall\julia\setup.jl:203
 [4] docall(call1::Ptr{Nothing})
   @ Main.JuliaCall C:\Users\domma\AppData\Local\R\win-library\4.2\JuliaCall\julia\setup.jl:176
```

**Reply by mcanouil** [reply]
And did you try the reverse, i.e., calling R from Julia via Julia Jupyter kernel?

Here the issue is on Kernel's side.
Quarto is only using them.

**Reply by DominiqueMakowski** [reply]
Interestingly, it works in Rstudio (access from both sides) but not in VScode. Any thoughts where to ask for help specific to quarto/VScode? Thanks a lot

**Reply by mcanouil** [reply]
There is no obvious reason for why it's not working in VSCode.
It's CLI, so the IDE does not matter.

Could you share a small "working" (reproducible) example to work with, _i.e._, a complete Quarto document or a Git repository? Thanks.
Also could you share the results of `quarto check`?

You can share a Quarto document using the following syntax, _i.e._, using more backticks than you have in your document (usually four ` ```` `).

`````md
````qmd
---
title: "Reproducible Quarto Document"
format: html
---

This is a reproducible Quarto document using `format: html`.
It is written in Markdown and contains embedded R code.
When you run the code, it will produce a plot.

```{r}
plot(cars)
```

The end.
````
`````

**Reply by DominiqueMakowski** [reply]
Interesting, here's a minimal quarto doc:

````qmd
---
title: "Reproducible Quarto Document"
format: html
---

```{julia}
x = 3
```

```{r}
JuliaCall::julia_eval("x")
```
````
![image](https://github.com/quarto-dev/quarto-cli/assets/8875533/4e433f5b-a2ba-43dc-842f-0ae8d4ab3658)

- From  RStudio
  - Rendering works ✅ 
  - Running the chunks from within in the console without rendering works ✅ 
- From VScode
  - Rendering works ✅ 

![image](https://github.com/quarto-dev/quarto-cli/assets/8875533/f5d564c1-6e47-42cc-84f8-95169f2334e9)


  - Running chunks doesn't 🔴 

![image](https://github.com/quarto-dev/quarto-cli/assets/8875533/cfa66381-2e39-4fec-9b4a-10f4de01901d)

Note that executing the code (CTRL + Enter) runs it in a Julia terminal (1.) and the object then shows in the Julia environment (2.). And running R code happens in the R terminal. Not sure if that's important

**Reply by mcanouil** [reply]
Yes, because when you run code cells, you are not using kernel/engine.
The extension guess to what terminal the code should be sent.

RStudio is focussed on R, thus everything runs through it.
VSCode is language agnostic, thus Julia goes to Julia, R to R, etc.

In conclusion, everything works as expected.

**Reply by DominiqueMakowski** [reply]
That's good to know, but is there a way to make it work in VScode? For instance by setting it up in a way that sends Julia code to the R console... assuming that would work

**Reply by mcanouil** [reply]
Change your keyboard shortcuts to execute the command you want.
